# <center> Predicting the Class of Mushroom with Logistic Regression </center>
#### *<center>By Peter Kayode</center>*

### Importation of neccessary Libraries

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import r2_score

In [2]:
mushroom_data = pd.read_csv("mushroom.csv")

mushroom_data

cap-shape cap-surface cap-color bruises%3F  odor gill-attachment  \
0         b'x'        b's'      b'n'       b't'  b'p'            b'f'   
1         b'x'        b's'      b'y'       b't'  b'a'            b'f'   
2         b'b'        b's'      b'w'       b't'  b'l'            b'f'   
3         b'x'        b'y'      b'w'       b't'  b'p'            b'f'   
4         b'x'        b's'      b'g'       b'f'  b'n'            b'f'   
...        ...         ...       ...        ...   ...             ...   
8119      b'k'        b's'      b'n'       b'f'  b'n'            b'a'   
8120      b'x'        b's'      b'n'       b'f'  b'n'            b'a'   
8121      b'f'        b's'      b'n'       b'f'  b'n'            b'a'   
8122      b'k'        b'y'      b'n'       b'f'  b'y'            b'f'   
8123      b'x'        b's'      b'n'       b'f'  b'n'            b'a'   

     gill-spacing gill-size gill-color stalk-shape  ...  \
0            b'c'      b'n'       b'k'        b'e'  ...   
1            b'c'      b'b'       b'k'        b'e'  ...   
2            b'c'      b'b'       b'n'        b'e'  ...   
3            b'c'      b'n'       b'n'        b'e'  ...   
4            b'w'      b'b'       b'k'        b't'  ...   
...           ...       ...        ...         ...  ...   
8119         b'c'      b'b'       b'y'        b'e'  ...   
8120         b'c'      b'b'       b'y'        b'e'  ...   
8121         b'c'      b'b'       b'n'        b'e'  ...   
8122         b'c'      b'n'       b'b'        b't'  ...   
8123         b'c'      b'b'       b'y'        b'e'  ...   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      b'w'                   b'w'      b'p'       b'w'   
1                      b'w'                   b'w'      b'p'       b'w'   
2                      b'w'                   b'w'      b'p'       b'w'   
3                      b'w'                   b'w'      b'p'       b'w'   
4                      b'w'                   b'w'      b'p'       b'w'   
...                     ...                    ...       ...        ...   
8119                   b'o'                   b'o'      b'p'       b'o'   
8120                   b'o'                   b'o'      b'p'       b'n'   
8121                   b'o'                   b'o'      b'p'       b'o'   
8122                   b'w'                   b'w'      b'p'       b'w'   
8123                   b'o'                   b'o'      b'p'       b'o'   

     ring-number ring-type spore-print-color population habitat class  
0           b'o'      b'p'              b'k'       b's'    b'u'  b'p'  
1           b'o'      b'p'              b'n'       b'n'    b'g'  b'e'  
2           b'o'      b'p'              b'n'       b'n'    b'm'  b'e'  
3           b'o'      b'p'              b'k'       b's'    b'u'  b'p'  
4           b'o'      b'e'              b'n'       b'a'    b'g'  b'e'  
...          ...       ...               ...        ...     ...   ...  
8119        b'o'      b'p'              b'b'       b'c'    b'l'  b'e'  
8120        b'o'      b'p'              b'b'       b'v'    b'l'  b'e'  
8121        b'o'      b'p'              b'b'       b'c'    b'l'  b'e'  
8122        b'o'      b'e'              b'w'       b'v'    b'l'  b'p'  
8123        b'o'      b'p'              b'o'       b'c'    b'l'  b'e'  

[8124 rows x 23 columns]

### Preprocessing the Data

In [3]:
labelencoder = LabelEncoder()
for i in mushroom_data.columns:
    mushroom_data[i] = labelencoder.fit_transform(mushroom_data[i])

mushroom_data

cap-shape  cap-surface  cap-color  bruises%3F  odor  gill-attachment  \
0             5            2          4           1     6                1   
1             5            2          9           1     0                1   
2             0            2          8           1     3                1   
3             5            3          8           1     6                1   
4             5            2          3           0     5                1   
...         ...          ...        ...         ...   ...              ...   
8119          3            2          4           0     5                0   
8120          5            2          4           0     5                0   
8121          2            2          4           0     5                0   
8122          3            3          4           0     8                1   
8123          5            2          4           0     5                0   

      gill-spacing  gill-size  gill-color  stalk-shape  ...  \
0                0          1           4            0  ...   
1                0          0           4            0  ...   
2                0          0           5            0  ...   
3                0          1           5            0  ...   
4                1          0           4            1  ...   
...            ...        ...         ...          ...  ...   
8119             0          0          11            0  ...   
8120             0          0          11            0  ...   
8121             0          0           5            0  ...   
8122             0          1           0            1  ...   
8123             0          0          11            0  ...   

      stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  \
0                          7                       7          0           2   
1                          7                       7          0           2   
2                          7                       7          0           2   
3                          7                       7          0           2   
4                          7                       7          0           2   
...                      ...                     ...        ...         ...   
8119                       5                       5          0           1   
8120                       5                       5          0           0   
8121                       5                       5          0           1   
8122                       7                       7          0           2   
8123                       5                       5          0           1   

      ring-number  ring-type  spore-print-color  population  habitat  class  
0               1          4                  2           3        5      1  
1               1          4                  3           2        1      0  
2               1          4                  3           2        3      0  
3               1          4                  2           3        5      1  
4               1          0                  3           0        1      0  
...           ...        ...                ...         ...      ...    ...  
8119            1          4                  0           1        2      0  
8120            1          4                  0           4        2      0  
8121            1          4                  0           1        2      0  
8122            1          0                  7           4        2      1  
8123            1          4                  4           1        2      0  

[8124 rows x 23 columns]

### Splitting the data into Training and Testing data

In [4]:
x = mushroom_data.drop("class", axis = 1)
y = mushroom_data["class"]

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 42)

### Fitting the Logistic Model

In [5]:
logistic_model = LogisticRegression(max_iter = 500)
logistic_model.fit(x_train, y_train)

LogisticRegression(max_iter=500)

### Predicting using the Model

In [6]:
pred = logistic_model.predict(x_test)
pred

array([0, 1, 1, ..., 1, 0, 0])

### Evaluating the model using Score

In [7]:
logistic_model.score(x_test, y_test) 

0.9495488105004102

In [8]:
confusion_matrix(y_test, pred)

array([[1201,   56],
       [  67, 1114]], dtype=int64)

Out of the edible mushroom, we correctly measure 1201 as true and correctly measure 1114 as false, not edible.

We then label 68 as being edible that were actually poisonous. So we should decide not to use this model in the health domain